<launch>
  
  <arg name="gui" default="false" />
  <arg name="gpu" default="-1" />
  <arg name="arm" default="larm" />
  <!-- <arg name="input_cloud" value="/kinect_head/depth_registered/points" /> -->
  <!-- <arg name="input_cloud" value="/kinect_head/depth_registered/quater/points" /> -->
  <arg name="input_base_scan" value="/base_scan_filtered" />
  <arg name="input_tilt_scan" value="/tilt_scan_filtered/navigation" />

  <arg name="input_camera_info" value="/kinect_head/rgb/camera_info" />
  <arg name="input_depth_image" value="/kinect_head/depth_registered/hw_registered/image_rect" />
  <arg name="input_pose_array" value="/edgetpu_human_pose_estimator/output/poses" />

  <arg name="manager_name" default="hand_navigation_pr2_manager" />
  <arg name="manager" default="/$(arg manager_name)" />

  <node name="$(arg manager_name)"
        pkg="nodelet" type="nodelet"
        args="manager"/>

  <node name="set_params"
        pkg="teach_spot" type="set_params_pr2.py"
        output="screen" />

  <!-- <node pkg="roseus" type="roseus" name="change_footprint" -->
  <!--       args="$(find teach_spot)/euslisp/pr2-footprint.l" > -->
  <!--   <param name="arm" value="$(arg arm)"/> -->
  <!-- </node> -->

  <node name="dr_spaam_lidar_person_detection"
        pkg="jsk_perception" type="lidar_person_detection_node.py"
        output="screen">
    <remap from="~input" to="/base_scan" />
    <rosparam subst_value="true" >
      map_link: /map
      weight_file: "$(find jsk_perception)/trained_data/lidar_person_detection/ckpt_jrdb_ann_dr_spaam_e20.pth"
      detector_model: "DR-SPAAM"
      conf_thresh: 0.8
      stride: 1
      panoramic_scan: false
      gpu: $(arg gpu)
    </rosparam>
  </node>

  <!-- <node name="people_pose_2d_to_3d" -->
  <!--       pkg="teach_spot" type="people_pose_2d_to_3d.py" > -->
  <!--   <remap from="~input/pose" to="$(arg input_pose_array)" /> -->
  <!--   <remap from="~input/info" to="$(arg input_camera_info)" /> -->
  <!--   <remap from="~input/depth" to="$(arg input_depth_image)" /> -->
  <!--   <rosparam subst_value="true"> -->
  <!--     approximate_sync: true -->
  <!--     queue_size: 100 -->
  <!--   </rosparam> -->
  <!-- </node> -->

  <node name="person_pose_array_to_mask" output="screen"
        pkg="teach_spot" type="person_pose_array_to_mask.py">
    <remap from="~input" to="/dr_spaam_lidar_person_detection/output" />
    <remap from="~input/info" to="/virtual_camera_mono/camera_info" />

    <rosparam>
      target_arm: $(arg arm)
      padding: 0.2
    </rosparam>
  </node>

  <node name="person_pose_array_to_boxes"
        pkg="teach_spot" type="person_pose_array_to_boxes.py">
    <remap from="~input" to="/dr_spaam_lidar_person_detection/output" />
    <rosparam>
      target_arm: $(arg arm)
      base_frame_id: base_link
    </rosparam>
  </node>

  <group ns="filter_base_laser_scan" >
    <node name="attention_clipper"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/AttentionClipper $(arg manager)" >
      <remap from="~input/points" to="$(arg input_base_scan)" />
      <remap from="~input/box_array" to="/person_pose_array_to_boxes/output" />
      <rosparam>
        use_multiple_attention: true
        negative: true
        approximate_sync: true
        queue_size: 1000
      </rosparam>
    </node>

    <node name="extract_indices"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/ExtractIndices $(arg manager)">
      <remap from="~input" to="$(arg input_base_scan)" />
      <remap from="~indices" to="attention_clipper/output/point_indices" />
      <remap from="~output" to="/base_scan_filtered_non_human" />
      <rosparam>
        keep_organized: true
        approximate_sync: true
        queue_size: 100
      </rosparam>
    </node>

    <node name="pointcloud_mux"
          pkg="topic_tools" type="mux"
          args="$(arg input_base_scan)_mux /base_scan_filtered_non_human $(arg input_base_scan)" >
      <remap from="mux" to="pointcloud_mux"/>
    </node>
  </group>

  <group ns="filter_tilt_laser_scan" >
    <node name="attention_clipper"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/AttentionClipper $(arg manager)" >
      <remap from="~input/points" to="$(arg input_tilt_scan)" />
      <remap from="~input/box_array" to="/person_pose_array_to_boxes/output" />
      <rosparam>
        use_multiple_attention: true
        negative: true
        approximate_sync: true
        queue_size: 1000
      </rosparam>
    </node>

    <node name="extract_indices"
          pkg="nodelet" type="nodelet"
          args="load jsk_pcl/ExtractIndices $(arg manager)">
      <remap from="~input" to="$(arg input_tilt_scan)" />
      <remap from="~indices" to="attention_clipper/output/point_indices" />
      <remap from="~output" to="/tilt_scan_filtered_non_human" />
      <rosparam>
        keep_organized: true
        approximate_sync: true
        queue_size: 100
      </rosparam>
    </node>

    <node name="pointcloud_mux"
          pkg="topic_tools" type="mux"
          args="/tilt_scan_filtered_mux /tilt_scan_filtered_non_human $(arg input_tilt_scan)" >
      <remap from="mux" to="pointcloud_mux"/>
    </node>
  </group>

  <group if="$(arg gui)" >
    <node name="$(anon rviz)"
          pkg="rviz" type="rviz"
          args="-d $(find teach_spot)/launch/config/human_filter.rviz" >
    </node>
  </group>
</launch>

